### PR TITLE
feat(core): add package manager parsers and stringifiers

### PR DESCRIPTION
**Pull request by meeroslav**
This PR adds `parseLockFile` and `writeLockFile` functions to `nx` necessary for future pruning functionality and dependency tracking during tasks.

## Related Issue(s)

Related to https://github.com/nrwl/nx/issues/9761

### CODE REVIEW

File: scripts/check-lock-files.js [hunk 1]
```diff
@@ -24,7 +24,7 @@ function checkLockFiles() {
   }
   try {
     require('child_process').execSync(
-      'yarn lockfile-lint -s -n -p yarn.lock -a hosts yarn npm',
+      'yarn lockfile-lint -s -p yarn.lock -a hosts yarn npm',
```
Comment (meeroslav) [inline scripts/check-lock-files.js:27]: We disable this explicitly, because of `pnpm`'s package name hacks.

**Comment by vsavkin** [comment]
I think the pair of parseLockFIle/writeLockFile is a good interface. 

Some notes and questions:

1. Should parseLockFIle be generic?
LockFileData is generic, but that type parameter is not used in the parseLockFile/writeLockFile functions, and I don't think we would use it.

After all, we want to write lock-file-oriented code in a polymorphic way. E.g., if we want to prune the lock file, we would want to write that code without any specific yarn, npm etc stuff. If it is possible, the parse/write functions cannot be generic. If it is not possible, then I don't think hose functions are needed because the if statement you have inside will have to outside. Does it make sense?

We use generics in the same way when creating a project graph. The generic parameter there is not used.

2. Should parseLockFIle take parameters?
 
 I don't think it should cause I imagine all the unit testing will be done around package-specific functions (you will test npm,yarn,pnpm versions directly).

3. I don't think the tests should be in the lockfile-test folder. All unit tests should be based on parsing and serializing strings. We need at least some unit tests for each package manager.

**Review by vsavkin — CHANGES_REQUESTED** [review]
Left some comments.

**Comment by meeroslav** [comment]
Thank you for checking the PR.

I'll check if we can get rid of generics without making code too messy. But generally, all callers of parse, stringify (and future prune) should be agnostic of package manager (and lock file generic type).

Regarding the lock files in script folder, all of that will be gone. It was more convenient while debugging, but I'll drop that entire folder and add tests with fixtures. This would also make the working directory param obsolete.

Once I add tests, the PR will be ready for full review.

**Comment by Nikamura** [comment]
Hey guys, why was `@zkochan/js-yaml` added as dependency instead of using `js-yaml` ?

https://www.npmjs.com/package/@zkochan/js-yaml/v/0.0.6
https://www.npmjs.com/package/js-yaml

**Comment by meeroslav** [comment]
@Nikamura  Because of the lock file pruning. The `@zkochan/js-yaml` adds additional functionality on top of `js-yaml`, which unfortunately is necessary to properly construct `pnpm-lock.yaml`.

**Pull request merged**